From 5c7a88e707e99922bc38eda792b2ff6c929b764f Mon Sep 17 00:00:00 2001
From: Clive Messer <clive.messer@digitaldreamtime.co.uk>
Date: Tue, 17 May 2016 08:39:12 +0000
Subject: [PATCH 45/65] ASoC: hw-params-rules: match_rate_{gt,lt,div_by,noop}

Add asoc_generic_hw_params_match_rate_{gt,lt,dev_by,noop} methods.

Signed-off-by: DigitalDreamtime <clive.messer@digitaldreamtime.co.uk>
---
 sound/soc/generic/hw-params-rules.c | 44 +++++++++++++++++++++++++++++++++++++
 1 file changed, 44 insertions(+)

diff --git a/sound/soc/generic/hw-params-rules.c b/sound/soc/generic/hw-params-rules.c
index c21a55e..5a9bae5 100644
--- a/sound/soc/generic/hw-params-rules.c
+++ b/sound/soc/generic/hw-params-rules.c
@@ -127,6 +127,44 @@ static int asoc_generic_hw_params_match_rate(
 	return 0;
 }
 
+static int asoc_generic_hw_params_match_rate_gt(
+	struct snd_pcm_substream *substream,
+	struct snd_pcm_hw_params *params,
+	void *data)
+{
+	long int rate = params_rate(params);
+
+	return (rate > (long int) data);
+}
+
+static int asoc_generic_hw_params_match_rate_lt(
+	struct snd_pcm_substream *substream,
+	struct snd_pcm_hw_params *params,
+	void *data)
+{
+	long int rate = params_rate(params);
+
+	return (rate < (long int) data);
+}
+
+static int asoc_generic_hw_params_match_rate_div_by(
+	struct snd_pcm_substream *substream,
+	struct snd_pcm_hw_params *params,
+	void *data)
+{
+	long int rate = params_rate(params);
+
+	return (rate % (long int) data == 0);
+}
+
+static int asoc_generic_hw_params_match_noop(
+	struct snd_pcm_substream *substream,
+	struct snd_pcm_hw_params *params,
+	void *data)
+{
+	return 1;
+}
+
 static int asoc_generic_hw_params_set_fixed_bclk_size(
 	struct snd_pcm_substream *substream,
 	struct snd_pcm_hw_params *params,
@@ -153,12 +191,18 @@ struct asoc_generic_hw_params_method {
 	HW_PARAMS_METHOD(n, asoc_generic_hw_params_read_u32)
 #define HW_PARAMS_METHOD_U32ARRAY(n) \
 	HW_PARAMS_METHOD(n, asoc_generic_hw_params_read_u32array)
+#define HW_PARAMS_METHOD_NOOP(n) \
+	HW_PARAMS_METHOD(n, NULL)
 
 static const struct asoc_generic_hw_params_method
 asoc_generic_hw_params_methods[] = {
 	HW_PARAMS_METHOD_U32ARRAY(asoc_generic_hw_params_match_sample_bits),
 	HW_PARAMS_METHOD_U32ARRAY(asoc_generic_hw_params_match_rate),
+	HW_PARAMS_METHOD_U32(asoc_generic_hw_params_match_rate_gt),
+	HW_PARAMS_METHOD_U32(asoc_generic_hw_params_match_rate_lt),
+	HW_PARAMS_METHOD_U32(asoc_generic_hw_params_match_rate_div_by),
 	HW_PARAMS_METHOD_U32ARRAY(asoc_generic_hw_params_match_channels),
+	HW_PARAMS_METHOD_NOOP(asoc_generic_hw_params_match_noop),
 	HW_PARAMS_METHOD_U32(asoc_generic_hw_params_set_fixed_bclk_size)
 };
 
-- 
2.7.4

